From 7c195c66cb6992f14552b979fee07a1100030be1 Mon Sep 17 00:00:00 2001
From: Yannick Serafini <yannick.serafini@wifx.net>
Date: Wed, 25 Feb 2026 22:47:16 +0100
Subject: [PATCH 06/12] usb: gadget: atmel_usba_udc: add basic USB role-switch
 support

Add initial USB role-switch support to atmel_usba_udc and wire it into
the existing UDC lifecycle.

Introduce role-switch registration with set/get callbacks and add basic
role application paths for none/device. Host requests are mapped to a
detached UDC path at this stage.

Gate pullup and UDC start paths so the device side is only activated
when the current role is device, and serialize role/VBUS transitions
through a shared state lock.

When role-switch is enabled, ignore VBUS GPIO handling to keep a single
source of truth for cable role control.

Signed-off-by: Yannick Serafini <yannick.serafini@wifx.net>
---
 .../devicetree/bindings/usb/atmel-usb.txt     |  16 ++
 drivers/usb/gadget/udc/atmel_usba_udc.c       | 215 ++++++++++++++++--
 drivers/usb/gadget/udc/atmel_usba_udc.h       |  10 +-
 3 files changed, 219 insertions(+), 22 deletions(-)

diff --git a/Documentation/devicetree/bindings/usb/atmel-usb.txt b/Documentation/devicetree/bindings/usb/atmel-usb.txt
index 12183ef47ee4..77062a753388 100644
--- a/Documentation/devicetree/bindings/usb/atmel-usb.txt
+++ b/Documentation/devicetree/bindings/usb/atmel-usb.txt
@@ -103,6 +103,10 @@ Deprecated property:
 Optional properties:
  - atmel,vbus-gpio: If present, specifies a gpio that allows to detect whether
    vbus is present (USB is connected).
+ - usb-role-switch: If present, registers the UDC as a USB role-switch
+	 provider and allows selecting USB_ROLE_NONE / USB_ROLE_DEVICE / USB_ROLE_HOST.
+	 In this mode, role selection is driven by the role-switch framework.
+	 This property must not be used together with atmel,vbus-gpio.
 
 Deprecated child node properties:
  - name: Name of the endpoint.
@@ -123,3 +127,15 @@ usb2: gadget@fff78000 {
 	clock-names = "hclk", "pclk";
 	atmel,vbus-gpio = <&pioB 19 0>;
 };
+
+usb2: gadget@fff78000 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	compatible = "atmel,at91sam9rl-udc";
+	reg = <0x00600000 0x80000
+	       0xfff78000 0x400>;
+	interrupts = <27 4 0>;
+	clocks = <&utmi>, <&udphs_clk>;
+	clock-names = "hclk", "pclk";
+	usb-role-switch;
+};
diff --git a/drivers/usb/gadget/udc/atmel_usba_udc.c b/drivers/usb/gadget/udc/atmel_usba_udc.c
index ec6bfbc6c264..e94b50c1f21f 100644
--- a/drivers/usb/gadget/udc/atmel_usba_udc.c
+++ b/drivers/usb/gadget/udc/atmel_usba_udc.c
@@ -25,6 +25,7 @@
 #include <linux/of.h>
 #include <linux/irq.h>
 #include <linux/gpio/consumer.h>
+#include <linux/usb/role.h>
 
 #include "atmel_usba_udc.h"
 #define USBA_VBUS_IRQFLAGS (IRQF_ONESHOT \
@@ -1947,7 +1948,7 @@ static irqreturn_t usba_vbus_irq_thread(int irq, void *devid)
 	/* debounce */
 	udelay(10);
 
-	mutex_lock(&udc->vbus_mutex);
+	mutex_lock(&udc->state_lock);
 
 	vbus = vbus_is_present(udc);
 	if (vbus != udc->vbus_prev) {
@@ -1965,16 +1966,129 @@ static irqreturn_t usba_vbus_irq_thread(int irq, void *devid)
 		udc->vbus_prev = vbus;
 	}
 
-	mutex_unlock(&udc->vbus_mutex);
+	mutex_unlock(&udc->state_lock);
 	return IRQ_HANDLED;
 }
 
+#ifdef CONFIG_USB_ROLE_SWITCH
+static bool usba_get_mux_locked(struct usba_udc *udc)
+{
+	return udc->mux_is_host;
+}
+
+static void usba_set_mux_locked(struct usba_udc *udc, bool host)
+{
+	unsigned long flags;
+	u32 ctrl;
+
+	if (udc->mux_is_host == host)
+		return;
+
+	spin_lock_irqsave(&udc->lock, flags);
+	ctrl = usba_readl(udc, CTRL);
+	if (host)
+		ctrl &= ~USBA_EN_USBA;
+	else
+		ctrl |= USBA_EN_USBA;
+	usba_writel(udc, CTRL, ctrl);
+	udc->mux_is_host = host;
+	spin_unlock_irqrestore(&udc->lock, flags);
+}
+
+static void usba_apply_role_none_locked(struct usba_udc *udc, bool mux_host)
+{
+	udc->suspended = false;
+
+	if (udc->driver)
+		usb_gadget_disconnect(&udc->gadget);
+
+	usba_stop(udc);
+	phy_set_mode_ext(udc->phy, PHY_MODE_USB_DEVICE, 0);
+	usba_set_mux_locked(udc, mux_host);
+}
+
+static int usba_apply_role_device_locked(struct usba_udc *udc)
+{
+	int ret;
+	bool mux_is_host = usba_get_mux_locked(udc);
+
+	usba_set_mux_locked(udc, false);
+
+	if (!udc->driver)
+		return 0;
+
+	udc->suspended = false;
+
+	phy_set_mode_ext(udc->phy, PHY_MODE_USB_DEVICE, 1);
+	ret = usba_start(udc);
+	if (ret) {
+		phy_set_mode_ext(udc->phy, PHY_MODE_USB_DEVICE, 0);
+		usba_set_mux_locked(udc, mux_is_host);
+		return ret;
+	}
+
+	usb_gadget_connect(&udc->gadget);
+
+	return 0;
+}
+
+static int usba_role_switch_set(struct usb_role_switch *sw, enum usb_role role)
+{
+	struct usba_udc *udc = usb_role_switch_get_drvdata(sw);
+	enum usb_role prev_role;
+	int ret = 0;
+
+	mutex_lock(&udc->state_lock);
+
+	if (udc->vbus_pin) {
+		ret = -EPERM;
+		goto out;
+	}
+
+	prev_role = udc->role_sw_current;
+	if (role != prev_role) {
+		udc->role_sw_current = role;
+		switch (role) {
+		case USB_ROLE_DEVICE:
+			ret = usba_apply_role_device_locked(udc);
+			break;
+		case USB_ROLE_HOST:
+			usba_apply_role_none_locked(udc, true);
+			break;
+		case USB_ROLE_NONE:
+		default:
+			usba_apply_role_none_locked(udc, false);
+			break;
+		}
+	}
+
+out:
+	if (ret)
+		udc->role_sw_current = prev_role;
+	mutex_unlock(&udc->state_lock);
+	return ret;
+}
+
+static enum usb_role usba_role_switch_get(struct usb_role_switch *sw)
+{
+	struct usba_udc *udc = usb_role_switch_get_drvdata(sw);
+
+	return READ_ONCE(udc->role_sw_current);
+}
+#endif
+
 static int atmel_usba_pullup(struct usb_gadget *gadget, int is_on)
 {
 	struct usba_udc *udc = container_of(gadget, struct usba_udc, gadget);
 	unsigned long flags;
 	u32 ctrl, tst;
 
+#if IS_ENABLED(CONFIG_USB_ROLE_SWITCH)
+	if (udc->role_sw &&
+	    READ_ONCE(udc->role_sw_current) != USB_ROLE_DEVICE)
+		return -EOPNOTSUPP;
+#endif
+
 	spin_lock_irqsave(&udc->lock, flags);
 
 	tst = usba_readl(udc, TST);
@@ -2006,7 +2120,14 @@ static int atmel_usba_start(struct usb_gadget *gadget,
 	udc->driver = driver;
 	spin_unlock_irqrestore(&udc->lock, flags);
 
-	mutex_lock(&udc->vbus_mutex);
+	mutex_lock(&udc->state_lock);
+
+#if IS_ENABLED(CONFIG_USB_ROLE_SWITCH)
+	if (udc->role_sw && udc->role_sw_current != USB_ROLE_DEVICE) {
+		mutex_unlock(&udc->state_lock);
+		return 0;
+	}
+#endif
 
 	if (udc->vbus_pin)
 		enable_irq(gpiod_to_irq(udc->vbus_pin));
@@ -2020,14 +2141,14 @@ static int atmel_usba_start(struct usb_gadget *gadget,
 			goto err;
 	}
 
-	mutex_unlock(&udc->vbus_mutex);
+	mutex_unlock(&udc->state_lock);
 	return 0;
 
 err:
 	if (udc->vbus_pin)
 		disable_irq(gpiod_to_irq(udc->vbus_pin));
 
-	mutex_unlock(&udc->vbus_mutex);
+	mutex_unlock(&udc->state_lock);
 
 	spin_lock_irqsave(&udc->lock, flags);
 	udc->devstatus &= ~(1 << USB_DEVICE_SELF_POWERED);
@@ -2040,6 +2161,8 @@ static int atmel_usba_stop(struct usb_gadget *gadget)
 {
 	struct usba_udc *udc = container_of(gadget, struct usba_udc, gadget);
 
+	mutex_lock(&udc->state_lock);
+
 	if (udc->vbus_pin)
 		disable_irq(gpiod_to_irq(udc->vbus_pin));
 
@@ -2048,6 +2171,8 @@ static int atmel_usba_stop(struct usb_gadget *gadget)
 
 	udc->driver = NULL;
 
+	mutex_unlock(&udc->state_lock);
+
 	return 0;
 }
 
@@ -2348,10 +2473,12 @@ static int usba_udc_probe(struct platform_device *pdev)
 		return PTR_ERR(hclk);
 
 	spin_lock_init(&udc->lock);
-	mutex_init(&udc->vbus_mutex);
+	mutex_init(&udc->state_lock);
 	udc->pdev = pdev;
 	udc->pclk = pclk;
 	udc->hclk = hclk;
+	udc->role_sw_current = USB_ROLE_NONE;
+	udc->role_sw = NULL;
 
 	platform_set_drvdata(pdev, udc);
 
@@ -2363,6 +2490,9 @@ static int usba_udc_probe(struct platform_device *pdev)
 	}
 
 	usba_writel(udc, CTRL, USBA_DISABLE_MASK);
+#if IS_ENABLED(CONFIG_USB_ROLE_SWITCH)
+	udc->mux_is_host = false;
+#endif
 	clk_disable_unprepare(pclk);
 
 	udc->usba_ep = atmel_udc_of_init(pdev, udc);
@@ -2384,23 +2514,54 @@ static int usba_udc_probe(struct platform_device *pdev)
 	}
 	udc->irq = irq;
 
+
+#if IS_ENABLED(CONFIG_USB_ROLE_SWITCH)
+	if (pdev->dev.of_node &&
+	    of_property_read_bool(pdev->dev.of_node, "usb-role-switch")) {
+		struct usb_role_switch_desc role_sw_desc = { 0 };
+
+		role_sw_desc.fwnode = dev_fwnode(&pdev->dev);
+		role_sw_desc.driver_data = udc;
+		role_sw_desc.name = dev_name(&pdev->dev);
+		role_sw_desc.set = usba_role_switch_set;
+		role_sw_desc.get = usba_role_switch_get;
+#ifdef DEBUG_SYSFS
+		role_sw_desc.allow_userspace_control = true;
+#endif
+
+		udc->role_sw =
+			usb_role_switch_register(&pdev->dev, &role_sw_desc);
+		if (IS_ERR(udc->role_sw)) {
+			ret = PTR_ERR(udc->role_sw);
+			udc->role_sw = NULL;
+			return ret;
+		}
+	}
+#endif
+
 	if (udc->vbus_pin) {
-		irq_set_status_flags(gpiod_to_irq(udc->vbus_pin), IRQ_NOAUTOEN);
-		ret = devm_request_threaded_irq(&pdev->dev,
-				gpiod_to_irq(udc->vbus_pin), NULL,
-				usba_vbus_irq_thread, USBA_VBUS_IRQFLAGS,
-				"atmel_usba_udc", udc);
-		if (ret) {
+		if (udc->role_sw) {
+			dev_warn(&pdev->dev,
+				 "already registered as usb-role-switch, ignoring vbus gpio\n");
 			udc->vbus_pin = NULL;
-			dev_warn(&udc->pdev->dev,
-				 "failed to request vbus irq; "
-				 "assuming always on\n");
+		} else {
+			irq_set_status_flags(gpiod_to_irq(udc->vbus_pin), IRQ_NOAUTOEN);
+			ret = devm_request_threaded_irq(&pdev->dev,
+					gpiod_to_irq(udc->vbus_pin), NULL,
+					usba_vbus_irq_thread, USBA_VBUS_IRQFLAGS,
+					"atmel_usba_udc", udc);
+			if (ret) {
+				udc->vbus_pin = NULL;
+				dev_warn(&udc->pdev->dev,
+					 "failed to request vbus irq; "
+					 "assuming always on\n");
+			}
 		}
 	}
 
 	ret = usb_add_gadget_udc(&pdev->dev, &udc->gadget);
 	if (ret)
-		return ret;
+		goto out_role_sw;
 	device_init_wakeup(&pdev->dev, 1);
 
 	usba_init_debugfs(udc);
@@ -2408,6 +2569,13 @@ static int usba_udc_probe(struct platform_device *pdev)
 		usba_ep_init_debugfs(udc, &udc->usba_ep[i]);
 
 	return 0;
+
+out_role_sw:
+#if IS_ENABLED(CONFIG_USB_ROLE_SWITCH)
+	if (udc->role_sw)
+		usb_role_switch_unregister(udc->role_sw);
+#endif
+		return ret;
 }
 
 static void usba_udc_remove(struct platform_device *pdev)
@@ -2423,6 +2591,13 @@ static void usba_udc_remove(struct platform_device *pdev)
 	for (i = 1; i < udc->num_ep; i++)
 		usba_ep_cleanup_debugfs(&udc->usba_ep[i]);
 	usba_cleanup_debugfs(udc);
+
+#if IS_ENABLED(CONFIG_USB_ROLE_SWITCH)
+	if (udc->role_sw) {
+		usb_role_switch_unregister(udc->role_sw);
+		udc->role_sw = NULL;
+	}
+#endif
 }
 
 #ifdef CONFIG_PM_SLEEP
@@ -2434,7 +2609,7 @@ static int usba_udc_suspend(struct device *dev)
 	if (!udc->driver)
 		return 0;
 
-	mutex_lock(&udc->vbus_mutex);
+	mutex_lock(&udc->state_lock);
 
 	if (!device_may_wakeup(dev)) {
 		udc->suspended = false;
@@ -2455,7 +2630,7 @@ static int usba_udc_suspend(struct device *dev)
 	enable_irq_wake(udc->irq);
 
 out:
-	mutex_unlock(&udc->vbus_mutex);
+	mutex_unlock(&udc->state_lock);
 	return 0;
 }
 
@@ -2475,13 +2650,13 @@ static int usba_udc_resume(struct device *dev)
 	}
 
 	/* If Vbus is present, enable the controller and wait for reset */
-	mutex_lock(&udc->vbus_mutex);
+	mutex_lock(&udc->state_lock);
 	udc->vbus_prev = vbus_is_present(udc);
 	if (udc->vbus_prev) {
 		phy_set_mode_ext(udc->phy, PHY_MODE_USB_DEVICE, 1);
 		usba_start(udc);
 	}
-	mutex_unlock(&udc->vbus_mutex);
+	mutex_unlock(&udc->state_lock);
 
 	return 0;
 }
diff --git a/drivers/usb/gadget/udc/atmel_usba_udc.h b/drivers/usb/gadget/udc/atmel_usba_udc.h
index 58a5cff057a5..f3166672d257 100644
--- a/drivers/usb/gadget/udc/atmel_usba_udc.h
+++ b/drivers/usb/gadget/udc/atmel_usba_udc.h
@@ -9,6 +9,7 @@
 
 #include <linux/gpio/consumer.h>
 #include <linux/phy/phy.h>
+#include <linux/usb/role.h>
 
 /* USB register offsets */
 #define USBA_CTRL				0x0000
@@ -329,8 +330,8 @@ struct usba_udc {
 	/* Protect hw registers from concurrent modifications */
 	spinlock_t lock;
 
-	/* Mutex to prevent concurrent start or stop */
-	struct mutex vbus_mutex;
+	/* Serialize start/stop and role/VBUS transitions */
+	struct mutex state_lock;
 
 	void __iomem *regs;
 	void __iomem *fifo;
@@ -341,6 +342,8 @@ struct usba_udc {
 	const struct usba_udc_errata *errata;
 	int irq;
 	struct gpio_desc *vbus_pin;
+	struct usb_role_switch *role_sw;
+	enum usb_role role_sw_current;
 	int num_ep;
 	struct usba_fifo_cfg *fifo_cfg;
 	struct clk *pclk;
@@ -349,6 +352,9 @@ struct usba_udc {
 	bool bias_pulse_needed;
 	bool clocked;
 	bool suspended;
+#if IS_ENABLED(CONFIG_USB_ROLE_SWITCH)
+	bool mux_is_host;
+#endif
 	bool ep_prealloc;
 
 	u16 devstatus;
-- 
2.43.0

